### DISCUSSION TITLE
Create an HD wallet via js

### DISCUSSION THREAD

**Discussion by samelamin** [trustwallet/wallet-core#3504] [Q&A]
Hi All, im sure this feature is available but the js documentation isnt available from what i see

I would like to create a derived wallet for a particular chain based on the index
for example, given index `32` and a private key, generate an address on a target chain. evm,cosmos,solana,polkadot etc

Is this possible using the js package? 

This is how im doing it based on `wallet-core.d.ts`
Is this the correct way of doing it? If so, why would we need to define the cointype when that is already defined in the derivation path? 

Also i noticed the derivation path doesnt take the index into account, im getting different evm addresses but cosmos,sol and polkadot always returns the same address regardless of `index`

```
  const eth_derivation_path = `m/44'/60'/0'/0`;
  const cosmos_derivation_path = `m/44'/180'/0'/1`;
  const solana_derivation_path = `m/44'/501'/0'/2`;
  const polkadot_derivation_path = `m/44'/354'/0'/2`;

  const { CoinType, HexCoding, HDWallet, AnyAddress, Derivation,  DerivationPath } = core;

  const wallet = HDWallet.create(256, "");
  const mnemonic = wallet.mnemonic();

  const eth_key = wallet.getKey(CoinType.ethereum, eth_derivation_path);
  const pubKey1 = eth_key.getPublicKeySecp256k1(false);
  const ethAddress = AnyAddress.createWithPublicKey(pubKey1, CoinType.ethereum);
  console.log(`Get Ethereum address1: ${ethAddress.description()}`);

  const key2 = wallet.getKey(CoinType.cosmos, cosmos_derivation_path);
  const pubKey2 = key2.getPublicKeyEd25519();
  const comosAddress = AnyAddress.createWithPublicKey(pubKey2, CoinType.cosmos);
  console.log(`Get cosmos address: ${comosAddress.description()}`);

  const key3 = wallet.getKey(CoinType.solana, solana_derivation_path);
  const pubKey3 = key3.getPublicKeyEd25519();
  const solanaAddress = AnyAddress.createWithPublicKey(pubKey3, CoinType.solana);
  console.log(`Get solana address3: ${solanaAddress.description()}`);

  const key4 = wallet.getKey(CoinType.polkadot, polkadot_derivation_path);
  const pubKey4 = key4.getPublicKeyEd25519();
  const polkadotAddress = AnyAddress.createWithPublicKey(pubKey4, CoinType.polkadot);
  console.log(`Get polkadot address: ${polkadotAddress.description()}`);

```

**Comment by samelamin** [comment]
@Milerius  I saw you converted this into a discussion

can you help with this

**Comment by weixuefeng** [comment]
you should call `TWHDWalletGetPrivateKeyFromExtended`. use extend privateKey and derivationPath for more child private key.

**Comment by fukemy** [comment]
Hi, can someone help? I always get modules not found
```
const { CoinType, HexCoding, HDWallet,
    AnyAddress, Derivation, DerivationPath
} = require("@trustwallet/wallet-core/dist/src/wallet-core.d.ts");

```

**Reply by satoshiotomakan** [reply]
Hi @fukemy, please use the following import:
```js
import { initWasm, TW } from "@trustwallet/wallet-core";

initWasm().then((WalletCore) => {
  const { CoinType, HexCoding, HDWallet,
    AnyAddress, Derivation, DerivationPath } = WalletCore;

  // Continue here
});
```

**Reply by fukemy** [reply]
> Hi @fukemy, please use the following import:
> 
> ```js
> import { initWasm, TW } from "@trustwallet/wallet-core";
> 
> initWasm().then((WalletCore) => {
>   const { CoinType, HexCoding, HDWallet,
>     AnyAddress, Derivation, DerivationPath } = WalletCore;
> 
>   // Continue here
> });
> ```

Thanks you so much, I tested with NodeJS(because I tend to write the server to hold the user's wallet). But I got this error:

```
import { initWasm, TW } from "@trustwallet/wallet-core";

initWasm().then((WalletCore) => {
    const { CoinType, HexCoding, HDWallet,
        AnyAddress, Derivation, DerivationPath } = WalletCore;
    const mnemonic = "abandon ability able about above absent absorb abstract absurd abuse access accident";
    const wallet = HDWallet.createWithMnemonic(mnemonic, "");
    const address = wallet.getAddressForCoin(CoinType.bitcoin);
    console.log("Bitcoin Address:", address);
    // Continue here
});

```

`return a};ja=(a,b,c)=>{a=ma(a)?new URL(a):la.normalize(a);fs.readFile(a,function(d,e){d?c(d):b(e.buffer)})};1<process.argv.length&&(da=process.argv[1].replace(/\\/g,"/"));process.argv.slice(2);process.on("uncaughtException",function(a){throw a;});process.on("unhandledRejection",function(a){throw a;});g.inspect=function(){return"[Emscripten Module object]"}}else if(ha){if("object"==typeof process&&"function"===typeof require||"object"==typeof window||"function"==typeof importScripts)throw Error("not compiled for this environment (did you build to HTML and try to run it not on the web, or set ENVIRONMENT to something - like node - and run it someplace else - like on the web?)");`

So I want to ask: does this support NodeJS project? Or I need to use it's on something else like ReactJS....

**Reply by satoshiotomakan** [reply]
No, we don't compile WalletCore for NodeJS at the moment

**Reply by fukemy** [reply]
> No, we don't compile WalletCore for NodeJS at the moment

thanks u so much

**Reply by fukemy** [reply]
> No, we don't compile WalletCore for NodeJS at the moment

I tested with NextJS project, but got this error 
```
initWasp error RuntimeError: Aborted(both async and sync fetching of the wasm failed)
    GET http://localhost:3000/_next/static/chunks/app/wallet-core.wasm 404 (Not Found)
    wasm streaming compile failed](page.js:18 wasm streaming compile failed: TypeError: Failed to execute 'compile' on 'WebAssembly': HTTP status code is not ok)
    page.js:18 falling back to ArrayBuffer instantiation
    initWasp error RuntimeError: Aborted(both async and sync fetching of the wasm failed)
    at n (wallet-core.js:37:429)
    at Ua (wallet-core.js:39:149)
    at eval (wallet-core.js:40:238)
```

Can you help me please. Here is the code
```
useEffect(() => {
    initWasm().then((WalletCore) => {
      const { CoinType, HexCoding, HDWallet,
        AnyAddress, Derivation, DerivationPath } = WalletCore;
      const mnemonic = "abandon ability able about above absent absorb abstract absurd abuse access accident";
      const wallet = HDWallet.createWithMnemonic(mnemonic, "")
      const address = wallet.getAddressForCoin(CoinType.bitcoin)
      console.log("Bitcoin Address:", address)
    }).catch(error => {
      console.log('initWasp error', error)
    })
  }, [])
```

**Reply by satoshiotomakan** [reply]
I wish I could help, but I haven't worked with NextJS. Please take a look at this https://stackoverflow.com/questions/78186231/how-can-i-use-openjphjs-with-next-js

**Reply by fukemy** [reply]
> https://stackoverflow.com/questions/78186231/how-can-i-use-openjphjs-with-next-js

I make it work now, it's just problem from NextJS, nevermind, I still got other error, as mention here: https://github.com/trustwallet/wallet-core/issues/3992 🥹🥹

**Reply by KoxyG** [reply]
> > https://stackoverflow.com/questions/78186231/how-can-i-use-openjphjs-with-next-js
> 
> I make it work now, it's just problem from NextJS, nevermind, I still got other error, as mention here: #3992 🥹🥹

Hi @fukemy  how did you makee this wok?, "Wallet core using nextjs"

**Reply by fukemy** [reply]
> Hi @fukemy how d

too long time, I forgot this sorry bro
